packages/mds-utils/state-machine.ts

```diff
@@ -1,57 +1,146 @@
-    [VEHICLE_EVENTS.cancel_reservation]: VEHICLE_STATES.available
-  },
-  [VEHICLE_STATES.trip]: {
-    [VEHICLE_EVENTS.trip_leave]: VEHICLE_STATES.elsewhere,
-    [VEHICLE_EVENTS.trip_end]: VEHICLE_STATES.available
+    [VEHICLE_EVENTS.comms_lost]: [VEHICLE_STATES.unknown],
+    [VEHICLE_EVENTS.missing]: [VEHICLE_STATES.unknown],
+    [VEHICLE_EVENTS.reservation_cancel]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.trip_start]: [VEHICLE_STATES.on_trip],
+    [VEHICLE_EVENTS.unspecified]: [VEHICLE_STATES.available]
   },
-  [VEHICLE_STATES.unavailable]: {
-    [VEHICLE_EVENTS.service_start]: VEHICLE_STATES.available,
-    [VEHICLE_EVENTS.deregister]: VEHICLE_STATES.inactive,
-    [VEHICLE_EVENTS.agency_pick_up]: VEHICLE_STATES.removed,
-    [VEHICLE_EVENTS.provider_pick_up]: VEHICLE_STATES.removed
+  [VEHICLE_STATES.unknown]: {
+    [VEHICLE_EVENTS.agency_drop_off]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.agency_pick_up]: [VEHICLE_STATES.removed],
+    [VEHICLE_EVENTS.comms_restored]: [
+      VEHICLE_STATES.available,
+      VEHICLE_STATES.elsewhere,
+      VEHICLE_STATES.reserved,
+      VEHICLE_STATES.on_trip,
+      VEHICLE_STATES.non_operational
+    ],
+    [VEHICLE_EVENTS.decommissioned]: [VEHICLE_STATES.removed],
+    [VEHICLE_EVENTS.provider_drop_off]: [VEHICLE_STATES.available],
+    [VEHICLE_EVENTS.unspecified]: [VEHICLE_STATES.available, VEHICLE_STATES.removed]
   }
 }
 
-const getNextState = (currStatus: VEHICLE_STATE, nextEvent: VEHICLE_EVENT): VEHICLE_STATE | undefined => {
+const getNextStates = (currStatus: VEHICLE_STATE, nextEvent: VEHICLE_EVENT): VEHICLE_STATE[] | undefined => {
   return stateTransitionDict[currStatus]?.[nextEvent]
 }
 
-const generateTransitionLabel = (
-  status: VEHICLE_STATE,
-  nextStatus: VEHICLE_STATE,
-  transitionEvent: VEHICLE_EVENT
-) => {
+// Filter for all states that have this event as a valid exiting event
+function getValidPreviousStates(
+  event: VEHICLE_EVENT,
+  states: VEHICLE_STATE[] = Object.keys(VEHICLE_STATES) as VEHICLE_STATE[]
+) {
+  return states.filter(state => {
+    return Object.keys(stateTransitionDict[state]).includes(event)
+  })
+}
+
+function isEventValid(event: VehicleEvent) {
+  const { event_types } = event
+  const finalEventType: VEHICLE_EVENT = event_types[event_types.length - 1]
+  return EVENT_STATES_MAP[finalEventType].includes(event.vehicle_state as VEHICLE_STATE)
+}
+
+function isEventSequenceValid(eventA: VehicleEvent, eventB: VehicleEvent) {
+  let prevStates: VEHICLE_STATE[] = [eventA.vehicle_state]
+  for (const eventTypeB of eventB.event_types) {
+    const validPreviousStates = getValidPreviousStates(eventTypeB, prevStates)
+    if (validPreviousStates.length > 0) {
+      const nextStates = validPreviousStates.reduce((acc: VEHICLE_STATE[], state) => {
+        const possibleNextStates = getNextStates(state, eventTypeB)
+        if (possibleNextStates) {
+          return [...acc, ...possibleNextStates]
+        }
+        return acc
```

Comment:
Super nit, but you may be able to make this a tad cleaner by using a `??` operator
```suggestion
        const possibleNextStates = getNextStates(state, eventTypeB) ?? []
        return [...acc, ...possibleNextStates]
```